pkg/x509/x509.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package x509
+
+import (
+	"crypto"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+)
+
+var (
+	EKUOID             = asn1.ObjectIdentifier{2, 5, 29, 37}
+	EKUTimestampingOID = asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 3, 8}
+)
+
+// VerifyCertChain verifies that the certificate chain is valid for issuing
+// timestamping certificates. The chain should start with a leaf certificate,
+// followed by any number of intermediates, and end with the root certificate.
+func VerifyCertChain(certs []*x509.Certificate, signer crypto.Signer) error {
+	// chain must contain at least one CA certificate and a leaf certificate
+	if len(certs) < 2 {
+		return errors.New("certificate chain must contain at least two certificates")
+	}
+
+	roots := x509.NewCertPool()
+	roots.AddCert(certs[len(certs)-1])
+
+	intermediates := x509.NewCertPool()
+	if len(certs) > 2 {
+		for _, intermediate := range certs[1 : len(certs)-1] {
+			intermediates.AddCert(intermediate)
+		}
+	}
+
+	// Verify the certificate chain
+	opts := x509.VerifyOptions{
+		Roots:         roots,
+		Intermediates: intermediates,
+		KeyUsages: []x509.ExtKeyUsage{
+			x509.ExtKeyUsageTimeStamping,
+		},
+	}
+	if _, err := certs[0].Verify(opts); err != nil {
+		return err
+	}
+
+	// Verify that all certificates but the leaf are CA certificates
```

Comment:
does x509.Certificate.Verify also verify that the root was self-signed?

Reply:
No, and I actually reported this as a bug to golang, issue 51953. The response was that it's up to the caller to decide what is trusted as a root, but they did recognize it's a bug to not check that the root is a CA (as in, I can't pass a non-CA leaf as a trusted CA cert).